### PR TITLE
Adding the firrtl proto.

### PR DESCRIPTION
**Pull request by kevintownsend**
This is just the initial upload. This needs to be built by sbt or whatever the build system will be.

Primary for better space performance. However, other reasons exist. Protos seem like a good choice for a multi-language machine readable IR.

### CODE REVIEW

File: src/main/proto/firrtl.proto [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+      // Required.
+      bytes value = 1;
+      repeated Expression arg = 2;
+      // Required.
+      Expression clk = 3;
+      // Required.
+      Expression en = 4;
+    }
+
+    message InlineVerilog {
```
Comment (azidar) [inline src/main/proto/firrtl.proto]: InlineVerilog, Assert, and Cover haven't been added yet, so perhaps just comment them out with a TODO? Or add an issue to discuss these new constructs, and reference the issue in the comment?

File: src/main/proto/firrtl.proto [hunk 2]
```diff
@@ -0,0 +1,451 @@
+syntax = "proto3";
+
+package hardware.chips.chisel3;
```
Comment (azidar) [inline src/main/proto/firrtl.proto]: Should change package, maybe to "proto"? Not sure...

File: src/main/proto/firrtl.proto [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+
+option java_package = "hardware.chips.chisel3.firrtl";
```
Comment (azidar) [inline src/main/proto/firrtl.proto]: Change this too?

File: src/main/proto/firrtl.proto [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+    // Required.
+    string name = 1;
+  }
+
+  message Circuit {
+    repeated Module module = 1;
+    repeated Top top = 2;
+  }
+
+  message Module {
```
Comment (azidar) [inline src/main/proto/firrtl.proto]: If we want to match the scala ir, then Module -> DefModule, and UserModule -> Module.

File: src/main/proto/firrtl.proto [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+
+    // Required.
+    oneof module {
+      ExternalModule external_module = 1;
+      UserModule user_module = 2;
+    }
+  }
+
+  message Statement {
+    message Wire {
```
Comment (azidar) [inline src/main/proto/firrtl.proto]: Similarly, Wire -> DefWire, Register -> DefRegister, etc.

File: src/main/proto/firrtl.proto [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+
+    message SIntType {
+      Width width = 1;
+    }
+
+    message ClockType {
+      // Empty.
+    }
+
+    message RecordType {
```
Comment (azidar) [inline src/main/proto/firrtl.proto]: RecordType -> BundleType

File: src/main/proto/firrtl.proto [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+    message Mux {
+      // Required.
+      Expression condition = 1;
+      // Required.
+      Expression t_value = 2;
+      // Required.
+      Expression f_value = 3;
+    }
+
+    message ExtractBits {
```
Comment (azidar) [inline src/main/proto/firrtl.proto]: Not sure why this is separate from PrimOp?

Comment (kevintownsend) [reply src/main/proto/firrtl.proto]: Fine by me. I am not sure what was behind that decision. Similarly Mux, SubIndex, SubAccess could be changed to.

File: src/main/proto/firrtl.proto (hunk 6 above)
Comment (kevintownsend) [reply src/main/proto/firrtl.proto]: Done.

File: src/main/proto/firrtl.proto (hunk 3 above)
Comment (kevintownsend) [reply src/main/proto/firrtl.proto]: Done.

File: src/main/proto/firrtl.proto (hunk 2 above)
Comment (kevintownsend) [reply src/main/proto/firrtl.proto]: Done.

File: src/main/proto/firrtl.proto (hunk 1 above)
Comment (kevintownsend) [reply src/main/proto/firrtl.proto]: Deleting for now.

**Comment by kevintownsend** [comment]
Can you be explicit about what you want with a "serialization/deserialization test, as well as integrating with Driver."? Ideally I would like to keep PRs small and manageable.

File: src/main/proto/firrtl.proto (hunk 4 above)
Comment (azidar) [reply src/main/proto/firrtl.proto]: Oh, it stands for "Definition", but for better or for worse, its the name of the corresponding firrtl ir node (in firrtl/ir/IR.scala).

**Comment by azidar** [comment]
With the PR, I'd like to see the code actually running in a simple test. That means adding a test in src/test/proto, demonstrating the serialization/deserialization of test into the proto representation, and back.

**Pull request merged**